### PR TITLE
JP-3658: JWST Accommodation of the Ramp Fitting Likelihood Algorithm

### PR DESCRIPTION
**Pull request by kmacdonald-stsci**
Resolves [JP-3658](https://jira.stsci.edu/browse/JP-3658)

Closes #

This PR addresses the use of the likelihood algorithm with ramp fitting.  This updates the `spec` for the arguments of ramp fitting to allow for `LIKELY` as an algorithm option.  Also, the detector1 pipeline is modified to do some special handling needed when using the likelihood algorithm.  In particular, there data needs at least 4 groups per integration and, since the likelihood algorithm does its own jump detection, the normal pipeline jump detection needs to be skipped when using the likelihood algorithm.

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [ ] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [ ] added relevant milestone
- [ ] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

File: jwst/pipeline/calwebb_detector1.py
Comment (melanieclarke) [inline jwst/pipeline/calwebb_detector1.py:90]: Here too - check message formatting.

**Review by melanieclarke — CHANGES_REQUESTED** [review]
This needs to wait until the stcal PR is merged:
https://github.com/spacetelescope/stcal/pull/278

In the meantime, a few comments on the structure and log messages here.  Also, we'll need to update the documentation here, to match the changes in stcal and add the new option for the `algorithm` argument.

File: jwst/pipeline/calwebb_detector1.py
Comment (melanieclarke) [reply jwst/pipeline/calwebb_detector1.py:83]: Oh, sorry, I misunderstood -- I see what you're doing now, with inserting the previous fragment at the beginning of the message.

Still, this would be more efficiently composed as a single string e.g.
```
                log.warning(f"For the LIKELY ramp fit algorithm NGROUPS must be "
                            f"at least 4.  Input has NGROUPS = {ngroups}.")
                log.warning("Defaulting to OLS_C.  Also, the normal "
                            "pipeline jump detection algorithm will be used.")
```

File: CHANGES.rst
Comment (melanieclarke) [inline CHANGES.rst:63]: Update comment - jump is no longer skipped by default.

File: jwst/pipeline/calwebb_detector1.py
Comment (melanieclarke) [inline jwst/pipeline/calwebb_detector1.py:80]: Is there a reason to check the shape here instead of the metadata?

File: jwst/pipeline/calwebb_detector1.py
Comment (melanieclarke) [inline jwst/pipeline/calwebb_detector1.py:83]: Add another message here to make it clear that it's defaulting to OLS_C instead of LIKELY.

**Review by melanieclarke — COMMENTED** [review]
A few more small tidy-up comments

File: jwst/pipeline/calwebb_detector1.py
Comment (kmacdonald-stsci) [reply jwst/pipeline/calwebb_detector1.py:80]: The actual data shape is what matters, so I check the actual data shape.

**Comment by melanieclarke** [comment]
Looking better, thanks!  It looks like we still need documentation updates for the ramp fit step, though.

**Comment by melanieclarke** [comment]
Testing on some real data with ngroups=3, it looks like it warns and switches to OLS_C correctly when the step is called from the detector1 pipeline.  When ramp_fit is called directly, though, it reports `Using algorithm = LIKELY`, but then the step unexpectedly succeeds -- it looks like it's quietly defaulting to OLS (python) instead.  It should warn and default to OLS_C.

Sorry for the late realization, but since we are no longer turning off jump detection by default, perhaps this check should actually just be moved from the detector1 pipeline to the ramp fit step instead, so we can handle both calling conditions in the same place.

**Comment by melanieclarke** [comment]
@kmacdonald-stsci - coming back to this one, now that the stcal PR is close to ready.  Can we move the check for ngroups into the ramp_fit step and add some likelihood documentation?

**Comment by kmacdonald-stsci** [comment]
> @kmacdonald-stsci - coming back to this one, now that the stcal PR is close to ready. Can we move the check for ngroups into the ramp_fit step and add some likelihood documentation?

If you want the `NGROUPS` check moved to the step code, then it can probably just be moved to `STCAL`, so the only change for this PR will be the addition of the `LIKELY` option to the `algorithm` parameter.

**Comment by melanieclarke** [comment]
> If you want the `NGROUPS` check moved to the step code, then it can probably just be moved to `STCAL`, so the only change for this PR will be the addition of the `LIKELY` option to the `algorithm` parameter.

It looks to me like the `ramp_fit_step` interface in jwst reports the algorithm it's going to use before it calls the stcal code, so it would be nice to have the check there, before that happens.  We will also need the documentation updates on the jwst side.

File: docs/jwst/ramp_fitting/description.rst
Comment (melanieclarke) [inline docs/jwst/ramp_fitting/description.rst:60]: This reference is throwing an error in the documentation build

File: jwst/ramp_fitting/ramp_fit_step.py
Comment (melanieclarke) [inline jwst/ramp_fitting/ramp_fit_step.py:424]: This message is long - let's split it up into two messages and make them warnings.  Also, the f needs to go at the front of all the substrings to get the variables in there properly.

```suggestion
                log.warning(f"When selecting the LIKELY ramp fitting algorithm the"
                            f" ngroups needs to be a minimum of {LIKELY_MIN_NGROUPS},"
                            f" but ngroups = {ngroups}.")
                log.warning("Due to this, the ramp fitting algorithm"
                            " is being changed to OLS_C.")
```

File: docs/jwst/ramp_fitting/description.rst
Comment (kmacdonald-stsci) [reply docs/jwst/ramp_fitting/description.rst:60]: Shouldn't we expect this to fail, since the corresponding link doesn't exist yet?

Comment (tapastro) [reply docs/jwst/ramp_fitting/description.rst:60]: I think this should build once we merge the stcal PR and make that release.

File: jwst/ramp_fitting/ramp_fit_step.py
Comment (melanieclarke) [reply jwst/ramp_fitting/ramp_fit_step.py]: Thanks for fixing the variable reference.  Can you please also add the `f` at the beginning of the substrings, so the variables are printed properly?  Also, split into two messages and make it a warning instead of an info.

**Comment by tapastro** [comment]
One (hopefully final!) request of this PR - can you update the pyproject.toml pins on stcal, and then merge the CHANGES conflict? stcal release coming very soon.
```
"stcal>=1.9.0,<2.0.0",
```

**Comment by tapastro** [comment]
Made an edit to my pin suggestion - bad upper pin suggestion. Sorry!

**Review by melanieclarke — APPROVED** [review]

**Pull request merged**
